### PR TITLE
Parallel checkpoint execution

### PR DESCRIPTION
**Pull request by halfprice**
## Description 

Add back parallel checkpoint execution since we have MVCC now. (Undo the effect of #10546).

Benchmark result:

command: `cargo run --release --bin sui-single-node-benchmark -- --tx-count 500000 --component checkpoint-executor --checkpoint-size 5 --num-input-objects 2 no-move`

before `TPS=16310.020876826722`
after `TPS=17611.214821598394`

## Test Plan 

Existing tests.

---
If your changes are not user-facing and not a breaking change, you can skip the following section. Otherwise, please indicate what changed, and then add to the Release Notes section as highlighted during the release process.

### Type of Change (Check all that apply)

- [ ] protocol change
- [ ] user-visible impact
- [ ] breaking change for a client SDKs
- [ ] breaking change for FNs (FN binary must upgrade)
- [ ] breaking change for validators or node operators (must upgrade binaries)
- [ ] breaking change for on-chain data layout
- [ ] necessitate either a data wipe or data migration

### CODE REVIEW

**Comment by halfprice** [comment]
I mainly rely on the existing tests for the coverage. But any suggestions for adding more tests are welcome!

**Review by williampsmith — COMMENTED** [review]
Seems good, but a good test to exercise some edge cases might be something like

```
for f in `seq 1 20`; do RUST_LOG=sui=debug,error MSIM_TEST_SEED=9$f cargo simtest test_simulated_load_restarts test_simulated_load_with_reconfig --no-capture > /tmp/test_$f 2>&1 &; done
```
and check all the temp log files generated to ensure they all passed. The best test would be `test_simulated_load_reconfig_restarts` but last I heard that test is flaky (and disabled as a result). Perhaps you could also run that test against some seeds and see what happens? If it passes against this PR, we can re-enable, and if it doesn't, at least it might be worth looking at the failure case to make sure it doesn't surface something legitimate before merging.

Happy to stamp once test plan is updated

File: crates/sui-core/src/checkpoints/checkpoint_executor/mod.rs [hunk 1]
```diff
@@ -985,6 +901,107 @@ fn get_unexecuted_transactions(
...
+                    "Transaction effects for effects digest {:?} do not exist in effects table",
+                    fx_digest
+                );
+            }
+            let fx = fx.unwrap();
+            (*fx.transaction_digest(), fx)
+        })
+        .collect();
+
+    for (tx, _) in &executable_txns {
```
[the remark this answers is not in the document]
Comment (halfprice) [reply crates/sui-core/src/checkpoints/checkpoint_executor/mod.rs:954]: We also need to access the tx here. Maybe we can first create a `shared_object_txns`, and use it to derive the `shared_effects_digests` and here (I think that requires save an extra copy of txns)?

**Comment by halfprice** [comment]
> Seems good, but a good test to exercise some edge cases might be something like
> 
> ```
> for f in `seq 1 20`; do RUST_LOG=sui=debug,error MSIM_TEST_SEED=9$f cargo simtest test_simulated_load_restarts test_simulated_load_with_reconfig --no-capture > /tmp/test_$f 2>&1 &; done
> ```
> 
> and check all the temp log files generated to ensure they all passed. The best test would be `test_simulated_load_reconfig_restarts` but last I heard that test is flaky (and disabled as a result). Perhaps you could also run that test against some seeds and see what happens? If it passes against this PR, we can re-enable, and if it doesn't, at least it might be worth looking at the failure case to make sure it doesn't surface something legitimate before merging.
> 
> Happy to stamp once test plan is updated

Thanks @williampsmith for the suggestion! Here are the test runs of the current code, which are all passed. Happy to run it again once we resolve the other comment.

Screenshot 2023-11-29 at 10 47 42 AM

**Review by williampsmith — APPROVED** [review]
Thanks! OOC did you try also enabling and running `test_simulated_load_reconfig_restart` to see if that succeeds now?

**Comment by halfprice** [comment]
> Thanks! OOC did you try also enabling and running `test_simulated_load_reconfig_restart` to see if that succeeds now?

Sorry I missed that. I just re-enabled this test and ran it 20 times and they all passed. Let me see if I can reproduce the flakiness on `main`. If not, I can re-enable the test.

File: crates/sui-core/src/checkpoints/checkpoint_executor/mod.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (williampsmith) [reply crates/sui-core/src/checkpoints/checkpoint_executor/mod.rs:968]: Oh gotcha missed that, feel free to leave as is

File: crates/sui-core/src/checkpoints/checkpoint_executor/mod.rs [hunk 2]
```diff
@@ -649,6 +514,59 @@ impl CheckpointExecutor {
...
+    authority_store: Arc<AuthorityStore>,
+    checkpoint_store: Arc<CheckpointStore>,
+    epoch_store: Arc<AuthorityPerEpochStore>,
+    transaction_manager: Arc<TransactionManager>,
+    accumulator: Arc<StateAccumulator>,
+    local_execution_timeout_sec: u64,
+    metrics: &Arc<CheckpointExecutorMetrics>,
+    data_ingestion_dir: Option<PathBuf>,
+) -> SuiResult {
+    debug!("Preparing checkpoint for execution",);
```
Comment (lxfind) [inline crates/sui-core/src/checkpoints/checkpoint_executor/mod.rs:531]: Would be good to at least include checkpoint number here

File: crates/sui-core/src/checkpoints/checkpoint_executor/mod.rs [hunk 3]
```diff
@@ -996,6 +914,109 @@ fn get_unexecuted_transactions(
...
+                    tx,
+                    digest_to_effects.get(tx.digest()).unwrap(),
+                    &authority_store,
+                )
+                .await?;
+        }
+    }
+
+    let exec_start = Instant::now();
+    let prepare_elapsed = exec_start - prepare_start;
```
Comment (lxfind) [inline crates/sui-core/src/checkpoints/checkpoint_executor/mod.rs:981]: You can use prepare_start.elapsed()

Comment (halfprice) [reply crates/sui-core/src/checkpoints/checkpoint_executor/mod.rs]: Good catch! Done.

**Comment by halfprice** [comment]
> LGTM. Another thing that you could do to make it cleaner latter, as far as I understand, is if you make the checkpoint number an instrument field up top on execute_checkpoint, maybe you can actually get rid of including checkpoint sequence number in all the debug logs.

Let me look into how to do that, and will send it out separately.

**Pull request merged**